"""
    Warning: This rule works on reads trimmed by quality score.
    To identify the fastq file who are trimmed and work without ambiguity, 
    you have to name your trimmed files: 
                - '{data_dir}/{data_dir}_trimmed.fq.gz' for unpaired reads

    Map reads for each sample with Bowtie2 require some parameters such as:
        -max_mismatches
        -bowtie_index
    These parameters have to be defined in the config file.
    You can add any other wanted parameters in the "other options" of the config file.
    
    This rule use samtools to convert sam file to bam file.
    The result file's name is '{data_dir}_bowtie2.bam'

"""


rule mapping:
        input: "results/trimmed/{samples_type}/{samples}_trimmed.fq.gz"
        output:"results/mapped/{samples_type}/{samples}_mapped.bam"
	params: index = config["bowtie2"]["index"], options = config["bowtie2"]["other_options"]#, qsub=config["qsub"] + " -q long -e {data_dir}/{data_dir}_bowtie_qsub.err -o {data_dir}/{data_dir}_bowtie_qsub.out
        message: """--- Mapping samples with Bowtie2."""
        shell: "bowtie2 {params.options} -q -x {params.index} -U {input} |samtools view -bS - > {output}"    
